Homework: Symfony 1

aurelijusb opened the pull request:
### Kaip atlikti užduotį?
* Padaryk [pull request](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/creating-a-pull-request) į šią repozitoriją ir atsiųskite nuorodą (paminint savo vardą) paštu.
(žr. :video_camera: [video paaiškinimą](https://www.youtube.com/watch?v=L9UACxeA4-M&feature=youtu.be))

### Užduotis
> Realybėje dažnai būna situacijos: `padaryk kaip šitas, tik truputi pakeisk ir... ant Symfony`

![Task summary](https://user-images.githubusercontent.com/1453957/68548948-40f9aa80-03fb-11ea-8797-8ace2cd73282.png)

Padaryti kuo panašiau į https://hw1.nfq2019.online naudojant Symfony karkasą (`šį branch`'ą), bet `Dešimt balų` turėtų rodyti paspaudus ant Jūsų vardo, o ne ant [Aurelijus](https://hw1.nfq2019.online/student?name=Aurelijus&project=hack%3Cb%3Eer%3C/b%3E%27is%20po%20.mySubdomain%20%26project%3D123)

### Vertinimo kriterijai

- [ ] Praeina testai (žr. :heavy_check_mark:, :x: prie commit'ų)
- [ ] SASS vs CSS (ar bandyta lokaliai paleisti WebPack/Encore)
- [ ] HTML Twig'e (bootstrap)
- [ ] PHP `foreach` (dudėtingesnėms transformacijoms)
- [ ] Twig `foreach` (nes kitaip neatvaizduosite masyvo)
- [ ] Twig'e `if`
- [ ] Twig'e nuorodos
- [ ] Gaunami naršyklės parametrai į Controller
- [ ] Pabandyta sukurti naują Controller (pvz. su `maker` bundle)
- [ ] Suvokiama Twig escape: nuorodoms
- [ ] Pasiūlymai kodo pagerinimui (_po deadline_)

### DUK

* **Ar CSS/JS bibliotekos keistos** Ne, naudojamas `kickstarter` _production` versija.
* **Iš kur gauti duomenis, negi rašyti ranka?** Pavyzdiniame puslapyje apačioje yra nuoroda `Duomenų failas`, nukreiptas į JSON failą: https://hw1.nfq2019.online/students.json (pvz. naršyklė atvaizduoja su _Save_ _Copy_ _Collapse_, bet `PHP` skaito `raw` versija :wink: )
* **Tai projektams naudoti tą patį pavadinimą, kur prie SSH?** Ne, žr. pavyzdį su `Education sharing platform` :vs: `devcollab`
* **Ar studentų sąrašo eilės tvakra svarbi**. Ne. Svarbiausia yra, kad būtų visi studentai (jei testai praiena...)

### Naudingos nuorodos
* Spalvos: https://getbootstrap.com/docs/4.0/getting-started/theming/#variable-defaults
* HTML: Boostrap: https://getbootstrap.com/
* TWIG: https://symfony.com/doc/current/templating.html
* Interface Kernel: https://symfony.com/doc/current/reference/configuration/kernel.html
* Maker bundle: https://symfony.com/doc/current/bundles/SymfonyMakerBundle/index.html
* Controller request: https://symfony.com/doc/current/controller.html#the-request-object-as-a-controller-argument
* PHP masyvai: https://www.php.net/manual/en/language.types.array.php
* PHP masyvų funkcijos: https://www.php.net/manual/en/ref.array.php
* PHP `foreach`: https://www.php.net/manual/en/control-structures.foreach.php

aurelijusb committed:
Building project and smoke tests

So it would be easier to test, if all parts are connected correctly.

Documentation:
 * https://help.github.com/en/actions/automating-your-workflow-with-github-actions
 * https://symfony.com/doc/current/routing.html#debugging-routes

aurelijusb committed:
Running end to end (Browser) tests

So it would be clear, that all parts are working correctly,
leaving more time to Code review.

Adding screenshots and debugging intermediate results,
so it would be faster to evaluate and/or fix issues.

Using Cypress as one of Headless browser based test framework.
Cypress is running browser as a separate process,
so it was needed for custom debug function to see intermediate results
in GitHub actions as well.

Using e2e-tests.sh to reach docker network used by docker-compose.yml.

Initial idea was to use GitHub Checks Annotations,
but Cypress already formats output with colors.
So using just different indentation.

Video recording disabled to not abuse GitHub –
it already does tremendous job.

Using GitHub Action "if: always()" to store screenshots for both
successful builds and errors.

Documentation:
 * https://www.cypress.io/
 * https://github.com/cypress-io/cypress/issues/3199
 * https://help.github.com/en/actions/automating-your-workflow-with-github-actions/development-tools-for-github-actions#set-a-warning-message-warning
 * https://help.github.com/en/actions/automating-your-workflow-with-github-actions/contexts-and-expression-syntax-for-github-actions#job-status-check-functions

aurelijusb committed:
Tests for Symfony 1 Homework

Intention is to create similar project in symfony,
as is deployed on https://hw1.nfq2019.online

Note: https://hw1.nfq2019.online/students.json can be useful.

aurelijusb committed:
Run tests on Pull requests

aurelijusb committed:
Run tests on Pull requests

aurelijusb committed:
More tests

To cover other edge cases.

aurelijusbanelis committed:
Git ignore for local Cypress

aurelijusbanelis committed:
Fix tests for students grouping

There was an error in https://hw1.nfq2019.online/
There should be 51 real students instead of 46 unique names.

aurelijusbanelis commented:
:exclamation: Namų darbų pavyzdžiuose buvo padaryta klaida – nebuvo rodomi studentai su tais pačiais vardais.
Todėl buvo rodomos klaidos:
![image](https://user-images.githubusercontent.com/11046422/68839225-eb5a2200-06c8-11ea-88e0-28bb60716825.png)
`AssertionError: Expected 46 elements of students: expected 51 to equal 46`

Todėl https://hw1.nfq2019.online/ versija **buvo pakeista**.
> Dar nespėjau atnaujinti jūsų Pull request'ų. Bet turėtų užtekti `merge` naujausia versiją iš šios repozitorijos.

**Vertinimui tai įtakos neturės.**
Jei dėl klaidos sugaišote papildomo laiko. Palengvinimas iš PHP pusės `Studentai` dalies sugeneravimui:
```php
   private function groupByStudents(array $projects)
    {
        $result = [];
        foreach ($projects as $projectName => $project) {
            foreach ($project['students'] as $student) {
                $result[] = ['student' => $student, 'project' => $projectName, 'mentors' => $project['mentors']];
            }
        }
        return $result;
    }
}
```

aurelijusb committed:
Infrastructure to mark common mistakes

Using example of Symfony Homework 1

Documentation:
 * https://help.github.com/en/actions/automating-your-workflow-with-github-actions/development-tools-for-github-actions#set-an-error-message-error
 * https://www.php.net/manual/en/class.recursivedirectoryiterator.php

aurelijusb committed:
Common mistakes: Additional escape filter not necessary

Auto esacepe is not a big issue (using warning instead of error),
but for learning it would be better to seek for higher quality.

Documentation:
 * https://help.github.com/en/actions/automating-your-workflow-with-github-actions/development-tools-for-github-actions#set-an-error-message-error
 * https://symfony.com/doc/4.3/templates.html#output-escaping

aurelijusbanelis committed:
Added another alternative to escape

aurelijusbanelis committed:
Do not define variables in Twig

It is better to separate concerts between preparation of business logic
(or complex computation) versus visualization of the data.

In practice "only few" define functions in Twig grows very fast
and maintaining Frontend become too hard.

If complex logic is useful only for displaying data,
Twig extensions are better option.

Documentation:
 * https://symfony.com/doc/current/templating/twig_extension.html
 * https://en.wikipedia.org/wiki/Separation_of_concerns
 * https://twig.symfony.com/doc/3.x/tags/set.html

aurelijusbanelis committed:
Add warning for Tech dept

Goal is to encourage attitude of clean code,
so there would be less conflicts during real day-to-day business.

Using it only as a warning (no failing tests) to separate
nice-to-have attitude and learning-flaws.

Documentation:
 * https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/about-pull-request-reviews
 * https://en.wikipedia.org/wiki/Technical_debt

aurelijusbanelis committed:
Use path command for all links

Because with project growing – it would be too hard to
find all hard-coded urls
and replace them with different (e.g. multi language/SEO) routes.

Documentation:
 * https://symfony.com/doc/4.2/templating.html#linking-to-pages

aurelijusbanelis committed:
More context for path and url_encode tests

So it would be easier to spot all errors and
the real cause of the error.

aurelijusbanelis committed:
Notice for default value usage.

So same comment (recommendation)
would not be needed to be written for every student.

aurelijusbanelis committed:
Fix tests

Cypress works in asynchronous way, but differently (couting last calls).

aurelijusbanelis committed:
Do not use external files as a main logic source

This makes automated testing a nightmare (unstable).
All main logic files should be within the system.

aurelijusbanelis committed:
Do not fail for url_encode vs path formatting

"/" symbol is formatted depending on Twig or PHP side is used to escape.
This does not change behavior on output of the content
(change is only in Browser's URL) – so false error.

aurelijusbanelis committed:
Requirement for SCSS variables

The beauty of CSS preprocessors is to extend existing frameworks,
not to write same plain CSS again and again.

Documentation:
 * https://getbootstrap.com/docs/4.0/getting-started/theming/#variable-defaults

aurelijusbanelis committed:
Requirement for SCSS variables

The beauty of CSS preprocessors is to extend existing frameworks,
not to write same plain CSS again and again.

Documentation:
 * https://getbootstrap.com/docs/4.0/getting-started/theming/#variable-defaults

aurelijusbanelis committed:
Requirement for SCSS variables

The beauty of CSS preprocessors is to extend existing frameworks,
not to write same plain CSS again and again.

Documentation:
 * https://getbootstrap.com/docs/4.0/getting-started/theming/#variable-defaults

aurelijusbanelis committed:
Improvent to path

aurelijusbanelis committed:
Do not urlencode in PHP

aurelijusbanelis committed:
Do not depend on external data

It will make testing harder (unstable).

aurelijusbanelis committed:
Notice about not using KernelInterface

aurelijusbanelis committed:
False positive for escape('url')

aurelijusb committed:
Fix false alarms and default parameter warnings

aurelijusb committed:
Do not allow global variables

Documentation:
 * https://symfony.com/doc/current/controller.html#the-request-object-as-a-controller-argument

aurelijusb committed:
Edge case for path

Also added common suggestion for more readable Twig array access.

aurelijusb committed:
Fix foreach condition

To not only check last line.
